MarkBind/markbind
# Add deploy -t/--travis to deploy via Travis

**What is the purpose of this pull request? (put "X" next to an item, remove the rest)**

• [X] Enhancement to an existing feature


Resolves #457.



**What is the rationale for this request?**
Currently, MarkBind supports deploying to GitHub Pages via `markbind deploy`. This is suitable for user documentation that only needs to be published with every new version. However, developer documentation is usually updated more frequently, and it will be useful to support a `-t/--travis` option to allow automatic deployment upon a push via Travis CI.

**What changes did you make? (Give an overview)**
I added a `-t/--travis [tokenVar]` option that can be run in Travis CI to automatically build and deploy a MarkBind site. `-t/--travis [tokenVar]` will:

1. Check that MarkBind is currently being run in Travis CI.
2. Retrieve the GitHub access token stored in `GITHUB_TOKEN` or a user specified `tokenVar` enviornment variable.
3. Extract the GitHub repo slug if the user has specified a repo in `site.json`, and construct a HTTPS URL containing the token for pushing changes.
4. Build and deploy the site.

**Is there anything you'd like reviewers to focus on?**
1. Currently, I assume that we are only deploying to GitHub. This implementation may need to change for other Git providers, specifically, the regex for extracting the repo slug and the authentication method (not sure if other Git providers do `https://[token]@provider/repo`) .

**Testing instructions:**
1. Create a MarkBind site on GitHub and add it to Travis CI. Sample instructions [here](https://deploy-preview-649--markbind-master.netlify.com/userguide/deployingthesite#deploying-to-git-hub-pages-via-travis-ci).
2. Add an appropriate `.travis.yml` to run `markbind deploy -t`. A sample `.travis.yml`:
(Differs from the one in the documentation as it does not restrict to only building from `master`.)
```yaml
language: node_js
node_js:
  - '8'
install:
  - npm i -g git://github.com/Xenonym/markbind.git#deploy-travis
script: markbind deploy --travis
```
3. Push some changes to trigger the Travis CI build. The site should deploy to GitHub Pages automatically.
